stream-chat-android-client/src/main/java/io/getstream/chat/android/client/events/caching/TypingEventCache.kt

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.client.events.caching
+
+import io.getstream.chat.android.client.events.ChatEvent
+import io.getstream.chat.android.client.events.TypingStartEvent
+import io.getstream.chat.android.client.events.TypingStopEvent
+import io.getstream.chat.android.client.events.caching.TypingEventKeyValue.Companion.toStartTypingEventKeyValue
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+/**
+ * Caches [TypingStartEvent]s and automatically "cleans" stale ones
+ * after [cleanStaleEventMs] has elapsed by sending out a [TypingStopEvent]
+ * counterpart of the original event.
+ *
+ * @param coroutineScope The coroutine scope used for timed cleaning
+ * of stale jobs.
+ * @param cleanStaleEventMs The time it takes before a [TypingStartEvent] is
+ * considered stale and should be "cleaned".
+ * @param onEventFired The lambda called after a typing event has been processed
+ * and cached if needed.
+ */
+internal class TypingEventCache(
+    private val coroutineScope: CoroutineScope = CoroutineScope(DispatcherProvider.IO),
+    private val cleanStaleEventMs: Long = EphemeralStartTypingEvent.DEFAULT_DELAY_TIME_MS,
+    private var onEventFired: (event: ChatEvent) -> Unit,
+) : ChatEventCache {
+
+    /**
+     *  A list of currently typing users. Only [EphemeralStartTypingEvent]s
+     *  are stored here, which are by design self "cleaning" [TypingStartEvent]s.
+     *
+     *  [TypingStopEvent] are not stored, but their processing removes their
+     *  [EphemeralStartTypingEvent] counterpart.
+     */
+    private val currentlyTypingUsers = mutableMapOf<TypingEventKeyValue, EphemeralStartTypingEvent>()
+
+    /**
+     * Processes the incoming chat event accordingly.
+     *
+     * @param event The chat event to be processed.
+     */
+    override fun processEvent(event: ChatEvent) {
+        if (event is TypingStartEvent) {
+            processTypingStartEvent(event)
+        } else if (event is TypingStopEvent) {
+            processTypingStopEvent(event)
+        }
```

Comment:
Good suggestion. I'll add an exception here because passing other events by mistake here will leave them unprocessed an our regression testing should catch any of those.